pkg/operator/operator.go

```diff
@@ -996,50 +1005,72 @@ func toStateErrors(err error) []*client.StateError {
 	return serrs
 }
 
-func generateRunReportFromTaskErrors(tge tasks.TaskGroupErrors) (runReport, string) {
-	failedTask := cmostr.ToPascalCase(tge[0].Name) + "Failed"
-	if len(tge) > 1 {
-		failedTask = "MultipleTasksFailed"
-	}
+func isUWMTaskErr(taskError tasks.TaskErr) bool {
+	return strings.HasPrefix(taskError.Name, UWMTaskPrefix)
+}
+
+// generateRunReportFromTaskErrors goes through the tasks errors and constructs a runReport
+// with the appropriate Degraded and Available conditions.
+func generateRunReportFromTaskErrors(tge tasks.TaskGroupErrors) runReport {
+	defaultReason := "MultipleTasksFailed"
+	var unavailableErrCount, degradedErrCount, unavailableUWMErrCount, degradedUWMErrCount int
 
-	degraded := &stateInfo{reason: failedTask, status: client.UnknownStatus}
-	available := &stateInfo{reason: failedTask, status: client.UnknownStatus}
+	degraded := &stateInfo{reason: defaultReason, status: client.UnknownStatus}
+	unavailable := &stateInfo{reason: defaultReason, status: client.UnknownStatus}
 
 	for _, terr := range tge {
 		// each task can return a single or multiple errors (as an Aggregate)
 		// each error can be a StateError or a generic error (fmt.Errorf)
 		for _, serr := range toStateErrors(terr.Err) {
 			switch serr.State {
 			case client.DegradedState:
-				degraded.messages = append(degraded.messages, serr.Reason)
+				degradedErrCount++
+				if isUWMTaskErr(terr) {
+					degradedUWMErrCount++
+				}
+				degraded.messages = append(degraded.messages, fmt.Sprintf("%s: %s", terr.Name, serr.Reason))
 				if !serr.Unknown {
 					degraded.status = client.TrueStatus
 				}
 
 			case client.UnavailableState:
-				available.messages = append(available.messages, serr.Reason)
+				unavailableErrCount++
+				if isUWMTaskErr(terr) {
+					unavailableUWMErrCount++
+				}
+				unavailable.messages = append(unavailable.messages, fmt.Sprintf("%s: %s", terr.Name, serr.Reason))
 				if !serr.Unknown {
-					available.status = client.FalseStatus
+					unavailable.status = client.FalseStatus
 				}
+			default:
+				klog.Errorf("StateError with an unsupported State: %s", serr.State)
 			}
 		}
 	}
 
-	rpt := runReport{}
-
-	if len(degraded.messages) != 0 {
-		rpt.degraded = degraded
-	} else {
-		rpt.degraded = asExpected(client.FalseStatus)
+	inferReason := func(uwmErrCount, totalErrCount int) string {
+		switch {
+		case len(tge) == 1:
+			return tge[0].Name + "Failed"
+		case uwmErrCount == totalErrCount:
+			return "UserWorkloadTasksFailed"
+		case uwmErrCount == 0:
+			return "PlatformTasksFailed"
+		default:
+			// If no errors to report, this will be ignored anyway
+			return defaultReason
+		}
```

Comment:
I find it tricky to follow all the implicit assumptions for this function. Purely for the sake of readability, what do you think about:
- Set a `defaultReason := tge[0].Name + "Failed"

and 
```suggestion
	inferReason := func(uwmErrCount, totalErrCount int) string {
		switch {
		case uwmErrCount == totalErrCount:
			return "UserWorkloadTasksFailed"
		case uwmErrCount == 0:
			return "PlatformTasksFailed"
	        case totalErrCount != 0:
	                return "MultipleTasksFailed"
		default:
			return defaultReason
		}
```

I think this requires less context (no need to remember what `tge` is.

Reply:
In case only one task is failing (len(tge)=1), we want to return its name, with the proposed change, this will override the "default" with UserWorkloadTasksFailed or PlatformTasksFailed

`one_failing_UWM_task_with_a_generic_error` case in tests e.g. will fail, you can run tests with `go test -v ./pkg/operator -run TestGenerateRunReportFromTaskErrors` (after removing the `No errors returned` case)

Reply:
ahh you are right of course. This _is_ tricky to understand and change :laughing: 
```suggestion
	inferReason := func(uwmErrCount, totalErrCount int) string {
		switch {
		case totalErrCount == 1:
			return defaultReason
		case uwmErrCount == totalErrCount:
			return "UserWorkloadTasksFailed"
		case uwmErrCount == 0:
			return "PlatformTasksFailed"
		default:
			return "MultipleTasksFailed"
		}
```
Might work?

But I'd also be happy with more comments to point out the assumptions that are baked in.

Reply:
This makes `multiple_failing_tasks_with_UWM_Degraded_StateError_and_Platform_Unavailable_StateError` fail,
As we can have a UWM degraded error and a Platform unavailable error. Now I remember, we don't keep track of the overall errors (only len(tge) can provide that, and I don't think adding a variable is worth it + what if ) but of the number of errors per state degraded or unavailable... I'll add this as comments + rename tge + maybe use some variables for readability...

 at least we're doing some mutation testing https://en.wikipedia.org/wiki/Mutation_testing and the tests are doing good :)

Reply:
I pushed some changes https://github.com/openshift/cluster-monitoring-operator/compare/be41464a356b83eb39d8d905df140773bad51a9f..c0e6d61567b7c59e9ccb553d3666205fc55ba3a8 tel me if it's ok

Reply:
ah the default reason is the issue with that :roll_eyes: gotcha. Ok lets go ahead with your additional change from 
> https://github.com/openshift/cluster-monitoring-operator/compare/be41464a356b83eb39d8d905df140773bad51a9f..c0e6d61567b7c59e9ccb553d3666205fc55ba3a8

I have a feeling this is not the last time we'll look at this code :)